### PR TITLE
Make path to downloaded binaries separately configurable

### PR DESCRIPTION
**Pull request by eu-smirnov**
Main changes:
- Move binary target handling logic from Storage class to package manager
- Add a configuration option for the path to downloaded files
- Use std::ifstream in public API instead of StorageRHandle

### CODE REVIEW

File: src/libaktualizr/primary/aktualizr_helpers.cc [hunk 1]
```diff
@@ -17,7 +17,7 @@ void targets_autoclean_cb(Aktualizr &aktualizr, const std::shared_ptr<event::Bas
     auto start = entry.installs.size() >= 2 ? entry.installs.end() - 2 : entry.installs.begin();
     for (auto it = start; it != entry.installs.end(); it++) {
       auto fit = std::find_if(installed_targets.begin(), installed_targets.end(),
-                              [&it](const Uptane::Target &t2) { return it->sha256Hash() == t2.sha256Hash(); });
+                              [&it](const Uptane::Target &t2) { return it->filename() == t2.filename(); });
```
Comment (patrickvacek) [inline src/libaktualizr/primary/aktualizr_helpers.cc:20]: Can we guarantee that filename will be unique? I'm wondering if we should match more than just that.

File: src/libaktualizr/primary/sotauptaneclient.h [hunk 2]
```diff
@@ -75,6 +75,16 @@ class SotaUptaneClient {
   void completeInstall() const;
   Uptane::LazyTargetsList allTargets() const;
   Uptane::Target getCurrent() const { return package_manager_->getCurrent(); }
+  std::vector<Uptane::Target> getStoredTargets() const { return package_manager_->getTargetFiles(); }
+  void deleteStoredTarget(const Uptane::Target &target) { package_manager_->removeTargetFile(target); }
+  std::ifstream openStoredTarget(const Uptane::Target &target) { 
+    auto status = package_manager_->verifyTarget(target);
```
Comment (patrickvacek) [inline src/libaktualizr/primary/sotauptaneclient.h:81]: +1

File: src/libaktualizr/storage/sqlstorage.cc [hunk 3]
```diff
@@ -1685,267 +1685,62 @@ void SQLStorage::clearDeviceData() {
...
-  boost::filesystem::path image_path_;
-  std::ifstream stream_;
-};
-
-std::unique_ptr<StorageTargetRHandle> SQLStorage::openTargetFile(const Uptane::Target& target) const {
-  return std_::make_unique<SQLTargetRHandle>(*this, target);
 }
 
-std::vector<Uptane::Target> SQLStorage::getTargetFiles() {
+std::vector<std::string> SQLStorage::getAllTargetnames() const {
```
Comment (patrickvacek) [inline src/libaktualizr/storage/sqlstorage.cc:1716]:
```suggestion
std::vector<std::string> SQLStorage::getAllTargetNames() const {
```

File: src/libaktualizr/storage/storage_common_test.cc [hunk 4]
```diff
@@ -571,7 +571,7 @@ TEST(storage, list_remove_targets) {
   EXPECT_EQ(tfs.size(), 0);
   EXPECT_FALSE(boost::filesystem::exists(temp_dir.Path() / "images" / "HASH"));
 }
-
+#endif
```
Comment (patrickvacek) [inline src/libaktualizr/storage/storage_common_test.cc:574]: Can these tests get moved elsewhere and restored?

File: src/libaktualizr/uptane/uptane_test.cc [hunk 5]
```diff
@@ -459,57 +461,58 @@ TEST(Uptane, InstallFakeBad) {
...
+  r.read(content, length);
+  EXPECT_EQ(r.gcount(), length);
+  r.close();
+  auto w = std::ofstream(image, std::ios::binary | std::ios::ate);
+  char content_bad[length + 1];
   memset(content_bad, 0, length + 1);
-  EXPECT_EQ(whandle->wfeed(content_bad, 3), 3);
-  whandle->wcommit();
+  w.write(content_bad, 3);
+  w.close();
```
Comment (patrickvacek) [inline src/libaktualizr/uptane/uptane_test.cc:478]: This test is such a hack but it does the job. :) It also conveniently tested a few of the old Storage APIs, though. Do we have good enough coverage of the new versions?

File: src/libaktualizr/primary/aktualizr_helpers.cc (hunk 1 above)
Comment (eu-smirnov) [reply src/libaktualizr/primary/aktualizr_helpers.cc:20]: That's the target name and I think there are a lot of places where we rely on it being unique.

File: src/libaktualizr/storage/storage_common_test.cc (hunk 4 above)
Comment (eu-smirnov) [reply src/libaktualizr/storage/storage_common_test.cc]: I've added two tests to packagemanagerfake, which cover the same functionality.

File: src/libaktualizr/uptane/uptane_test.cc (hunk 5 above)
Comment (eu-smirnov) [reply src/libaktualizr/uptane/uptane_test.cc:478]: It should be covered by new unit tests in packagemanagerfake and storage.

**Comment by eu-smirnov** [comment]
I've added a few unit tests to cover refactored methods and hopefully solved the remaining CI issues.

File: src/libaktualizr/package_manager/packagemanagerinterface.cc [hunk 6]
```diff
@@ -182,13 +180,97 @@ TargetStatus PackageManagerInterface::verifyTarget(const Uptane::Target& target)
...
+  if (!file) {
+    throw std::runtime_error("File doesn't exist for target " + target.filename());
+  }
+  boost::filesystem::remove(file->second);
+  storage_->deleteTargetInfo(target.filename());
+}
+
+std::vector<Uptane::Target> PackageManagerInterface::getTargetFiles() {
+  std::vector<Uptane::Target> v;
+  auto names = storage_->getAllTargetNames();
```
Comment (patrickvacek) [inline src/libaktualizr/package_manager/packagemanagerinterface.cc:270]: Why return just the names as strings from `getAllTargetNames` if we just construct fake Targets here?

Comment (eu-smirnov) [reply src/libaktualizr/package_manager/packagemanagerinterface.cc:270]: Do you mean why not construct fake targets in the storage method, or why construct them at all?

Comment (patrickvacek) [reply src/libaktualizr/package_manager/packagemanagerinterface.cc:270]: Both, I guess. It looks like previously we were constructing Targets in SqlStorage::getTargetFiles() based on the hashes that were stored in the database. What's the benefit of losing that information?

Comment (eu-smirnov) [reply src/libaktualizr/package_manager/packagemanagerinterface.cc:270]: There is no benefit in storing this information. Those Targets were still fake and used only for bookkeeping, as the comment in API header states:

>   Get list of targets currently in storage. This is intended to be used with
>   DeleteStoredTarget and targets are not guaranteed to be verified and
 >  up-to-date with current metadata.

For bookkeeping, it is sufficient to have only a filename.
I also think it is cleaner to keep in Storage only that logic that is directly related to reading/storing necessary information from db.

Comment (patrickvacek) [reply src/libaktualizr/package_manager/packagemanagerinterface.cc:270]: Fair enough, thanks for clarifying!

**Pull request merged**